src/main/kotlin/org/malachite/estella/services/ApplicationStageDataService.kt

```diff
@@ -59,18 +61,75 @@ class ApplicationStageDataService(
         }
 
 
-    fun setNotesToInterview(id: UUID, password: String, notes: Set<NotesFilePayload>) {
-        if (!canDevUpdate(id, password)) throw UnauthenticatedException()
-        val interview = interviewService.getInterview(id)
-        val savedNotes = interview.applicationStage.notes.plus(noteService.updateNotes(notes))
-        applicationStageRepository.save(interview.applicationStage.copy(notes = savedNotes))
-    }
+    fun setNotesToInterview(id: UUID, password: String?, notes: Set<NotesFilePayload>) =
+        interviewService.getInterview(id).applicationStage
+            .let { setNotesToApplicationStage(it, password, notes) }
+
+    fun setNotesToTaskStage(id: UUID, password: String?, notes: Set<NotesFilePayload>) =
+        taskStageService.getTaskStage(id).applicationStage.let { setNotesToApplicationStage(it, password, notes) }
+
+    fun setNotesToApplied(application: Application, password: String?, notes: Set<NotesFilePayload>) =
+        application.applicationStages.first().let { setNotesToApplicationStage(it, password, notes) }
+
+    private fun setNotesToApplicationStage(
+        applicationStage: ApplicationStageData,
+        password: String?,
+        notes: Set<NotesFilePayload>
+    ) = canAccessApplicationStageData(applicationStage, password)
+        .let { applicationStage.notes.plus(noteService.updateNotes(notes)) }
+        .let { applicationStageRepository.save(applicationStage.copy(notes = it)) }
+
+
+    fun getNotesByApplication(application: Application, password: String?): List<Note> =
+        application.applicationStages.first()
+            .let { getNotesByApplicationStage(it, password) }
+
+
+    fun getNotesByInterviewId(id: UUID, password: String?): List<Note> =
+        interviewService.getInterview(id).applicationStage
+            .let { getNotesByApplicationStage(it, password) }
+
+
+    fun getNotesByTaskId(id: UUID, password: String?): List<Note> =
+        taskStageService.getTaskStage(id).applicationStage
+            .let { getNotesByApplicationStage(it, password) }
+
+    private fun getNotesByApplicationStage(applicationStage: ApplicationStageData, password: String?): List<Note> =
+        canAccessApplicationStageData(applicationStage, password)
+            .let { applicationStage.application.applicationStages.flatMap { it.notes } }
+
+    fun getNotesByTaskIdWithTask(id: UUID, password: String?): TasksNotes =
+        taskStageService.getTaskStage(id).applicationStage
+            .let { getNotesWithTaskResults(it, password) }
+
+    private fun getNotesWithTaskResults(applicationStage: ApplicationStageData, password: String?): TasksNotes =
+        applicationStage
+            .also { canAccessApplicationStageData(it,password) }
+            .let { Pair(it.tasksStage!!.tasksResult, it.notes.toList()) }
+
+
+    private fun canAccessApplicationStageData(applicationStage: ApplicationStageData, password: String?) =
```

Comment:
To głupie jest trochę, metoda, która w nazwie ma `canAccess` zwraca `bool`, więc logika mówiłaby nam, że dostaniemy info czy może czy nie może, a okazuje się, że metoda zawsze zwraca `true`, oprócz momentu kiedy wali exceptionem. Można:
1. wywalić stąd rzucanie wyjątków, dać tam false (przy okazji zająć się `if(warunek) false else true`)
2. nic nie zwracać, walić wyjątek jeśli nie ma permsa, a nazwę zmienić na `assertAccess` czy jakieś `requireAccess`
Imo lepiej 1, ale zrozumiem, jeśli wygodniejsze będzie 2